src/empirical.jl

```diff
@@ -51,18 +51,43 @@ struct EmpiricalVariogram
     sums   = zeros(nlags)
     counts = zeros(Int, nlags)
 
+    # decide whether to construct a spatial index
+    # coordinates need to be AbstractFloat for NearestNeighbors to work
+    # https://github.com/KristofferC/NearestNeighbors.jl/issues/13
+    isfloat = eltype(coordinates(sdata)) <: AbstractFloat
+    # quite arbitrary, but for small datasets setting up the index is too much overhead
+    # it really also depends on what fraction of points are within hmax
+    isbig = npts > 1000
+
+    if isfloat && isbig
+      neigh = BallNeighborhood(hmax, distance)
+      neighsearcher = NeighborhoodSearcher(sdata, neigh)
+      use_spatial_index = true
+    else
+      use_spatial_index = false
+    end
+
     # preallocate memory for coordinates
     xi = MVector{ndims(sdata),coordtype(sdata)}(undef)
     xj = MVector{ndims(sdata),coordtype(sdata)}(undef)
 
-    @inbounds (
-    for j in 1:npts
+    @inbounds for j in 1:npts
       coordinates!(xj, sdata, j)
-      for i in j+1:npts
+
+      if use_spatial_index
```

Comment:
We could move this check that is happening inside the loop to outside. Actually we should refactor the code to look like two major branches: one branch in which we use the nearest neighbour search and another branch that doesn't. Could you refactor the code so that it looks more compact? Something like:

```julia
if is float && isbig
  # construct the variogram with neighbour search
else
  # construct the variogram with naive loops
end
```

Reply:
Yes, I refactored it to this structure in 0009ea940c63a829c0238850f89467d7e5bf8557. It does result in some more duplicated lines, if you have further ideas on how to structure it, feel free to also push to the branch.

Note that this check, even though it was inside the loop, was not expensive. I'm not sure, but it might have been optimised out completely.